### PR TITLE
Implement ephemeral containers subresource

### PR DESCRIPTION
**Pull request by jmintb**
## Motivation

Implement support for another sub resource, as described in #127.

## Solution

Support for the ephemeral containers subresource is implemented using a marker `trait` `Ephemeral`, which is implemented for the `Pod` struct. Three methods are implemented for `Api<K>` where `K` is bounded with the `Ephemeral` trait, `replace_ephemeral_containers`, `patch_ephemeral_containers` and `get_ephemeral_containers`. These methods correlate with the operations described in the [Kubernetes API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/), see the section on EphemeralContainers Operations.

### Questions

I have a few questions before this should be merged.

1. I noticed that the patch operation did not return errors for invalid patches. Is that expected behavior when patching a resource? See this [comment](https://github.com/kube-rs/kube/issues/127#issuecomment-1371162352) for examples.
2. This functionality was stabilized in K8S 1.25, so we need to guard this functionality.  A guard was mentioned [here](https://github.com/kube-rs/kube/issues/127#issuecomment-1297418940) but I am not sure how to apply it in this case. Are there other examples in kubers of guarding features based on K8S versions to use as inspiration?
3. I tried to follow the existing structure for sub resources hence the marker trait, but I am unsure if it makes sense in this case. Since ephemeral containers are only relevant for Pods. Do you think the marker trait makes sense here?

### CODE REVIEW

**Comment by clux** [comment]
Thanks a lot for this. Having had a quick look it looks generally sensible and fits in with the existing conventions.

> I have a few questions before this should be merged.

> 1. I noticed that the patch operation did not return errors for invalid patches. Is that expected behavior when patching a resource? See this [comment](https://github.com/kube-rs/kube/issues/127#issuecomment-1371162352) for examples.

hm.. yeah, that is unfortunate. that does not __feel__ like it's expected. maybe there's an upstream issue for it in kubernetes (or maybe no one has noticed it if everyone so far using it has come from client-go). it's another awkward kubernetes api where they expect you to supply one struct, but simultaneously nest it inside a bigger struct they won't use, so i'm sure there are interesting edge cases.

> 2. This functionality was stabilized in K8S 1.25, so we need to guard this functionality.  A guard was mentioned [here](https://github.com/kube-rs/kube/issues/127#issuecomment-1297418940) but I am not sure how to apply it in this case. Are there other examples in kubers of guarding features based on K8S versions to use as inspiration?

Ah, yeah, we have used the [k8s-openapi macros](https://docs.rs/k8s-openapi/latest/k8s_openapi/#macros) for this before. Here's the [last one we removed](https://github.com/kube-rs/kube/pull/1116/files#diff-d3995adb12082edc7a72e0976649a2de821047291f07eb884a69d8010ec512d2).

> 3. I tried to follow the existing structure for sub resources hence the marker trait, but I am unsure if it makes sense in this case. Since ephemeral containers are only relevant for Pods. Do you think the marker trait makes sense here?

Yeah, I think so. We have other marker traits that are only relevant to pods (like `Evict`), and it will become handy to have these traits if we ever get around to switching serialization format to [k8s-pb](https://github.com/kube-rs/k8s-pb).

**Review by clux — COMMENTED** [review]
some minor nits, interface tweaks, and test structure stuff.

the only thing here that i don't like is the api condition kubernetes is imposing on us (having to append to a vector). if there's any way to get around that through merge patches or server-side apply, then that would be a great thing to figure out as a better alternative for us to provide in the docs.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:189]: That makes more sense it felt very unergonomic. The other replace_* methods for sub resources do use Vec<u8> though, is the inconsistency alright?

File: kube-client/src/lib.rs
Comment (jmintb) [reply kube-client/src/lib.rs:655]: I might have gotten a bit carried away :sweat_smile:

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:250]: That looks good!, I will use the wrapper.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:168]: I hadn't tested the other patch varients :facepalm: , strategic merge and server side apply append the new container  thank god. I will update the docs to reflect that.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:230]: After a quick investigation I think this is caused by a Pod object happily being deserialized from a PodSpec json snippet. Where we actually end up sending an empty Pod object as the patch. I will perform more tests later, but essentially what I think happens is: 
1. we pass in a `Patch<P>`, where P is a PodSpec: https://github.com/kube-rs/kube/blob/4594a4f60b903b43d2f1873c2c353de24822dc25/kube-client/src/api/subresource.rs#L278
2. a request is built and the pod spec is serialized as the body, at this point we are stilling dealing with P: https://github.com/kube-rs/kube/blob/fdd799d4230aa09405881f18b946656637fed762/kube-core/src/request.rs#L259
3. Then before the actual request is performed the body is mapped to the return type of method which is Pod in this case, with this line, where I think we end up producing an empty Pod object as the body: https://github.com/kube-rs/kube/blob/75578903318a16688e14e30afa8745eee2c12f95/kube-client/src/client/mod.rs#L225.

Does this make sense? :thinking:

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs:189]: ugh, yes, those should also follow the same pattern ideally. i think those are safe to change in a non-breaking way as well though 🤔

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs:250]: Nice. Note I am doing the same in core_methods now (but not touching this file to avoid conflicts in this file) in #1158

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs:230]: hm, that should be easily testable with a println on the result there, but that seems unlikely to me.
the Request fn is like line-for-line equivalent to the `patch_metadata`.

i'm going to backtrack a bit here and say that this might not be inconsistent and "that bad" after all. types already need the wrapping in general, but we don't notice it for things like `status` because it's top level. noticed that it's also required for `patch_metadata` yesterday.

thus I think it's fine for the example to fully wrap in the json! patch:

```rust
serde_json::from_value(serde_json::json!({
    ///     "apiVersion": "v1",
    ///     "kind": "Pod",
    ///     "spec": {
    ///        "ephemeralContainers": [
    ///        {
    ///            "name": "myephemeralcontainer",
    ///            "image": "busybox:1.34.1",
    ///            "command": ["sh", "-c", "sleep 20"],
    ///        },
    ///     }
    /// }))?;
```

and just use server side apply in the example via `Patch::Apply`. If it's actually the same as `patch_metadata`, then you can omit the typemeta if you use a `Patch::Merge` (but including typemeta is valid in both cases)

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: I took the liberty of updating the other sub resource documentation examples as well.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: If we use `K`I think this would become a breaking change in the other methods as `K`is a resource.

This line `serde_json::to_vec(&json!(first_patch))?` would no longer work.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: Might have hidden too many imports in the examples though :thinking:

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:249]: > After a quick investigation I think this is caused by a Pod object happily being deserialized from a PodSpec json snippet. Where we actually end up sending an empty Pod object as the patch. I will perform more tests later, but essentially what I think happens is:

So my hypothesis was incorrect haha, I am not sure why no error is returned exactly other than perhaps Kubernetes ignoring all of the fields that do not belong to the resource being patched.

Just to be clear this is the case I am describing in the docs and brought up earlier.

```rust
        let patch: PodSpec = serde_json::from_value(json!({
           "spec" {
                 "ephemeralContainers:" [{
                    "name": "myephemeralcontainer2",
                    "image": "busybox:1.35.0",
                    "command": ["sh", "-c", "sleep 1"],
            }]
        }))?;

       pods.patch_ephemeral_containers(
                &pod_name,
                &PatchParams::default(),
                &Patch::Strategic(patch),
            )
            .await?
            .spec
            .unwrap()
            .ephemeral_containers
            .expect("could find ephemeral container");
```

No error will be returned here so it would be easy to run this and think everything is fine.

>i'm going to backtrack a bit here and say that this might not be inconsistent and "that bad" after all. types already need the wrapping in general, but we don't notice it for things like status because it's top level (and maybe also because we were using merge patches were it genuinely is not required).

I think the wrapping is just fine and consistent with how the rest of kube-rs and how client-go behaves as well. I am only worried about the lack of errors when submitting an invalid patch, hence the examples in the documentation. If you would prefer to just focus on the working example in the documentation and drop the valid/invalid comparisons I will remove them.

> and just use server side apply in the example via Patch::Apply. If it's actually the same as patch_metadata, then you can omit the typemeta if you use a Patch::Merge (but including typemeta is valid in both cases)

The example currently uses `Patch::Strategic` and no type metadata is required.

File: kube-client/src/api/mod.rs
Comment (clux) [reply kube-client/src/api/mod.rs:19]: hmm, you would at the very least need the same gate the integration tests in lib.rs as well because they'll not pass the MK8SV check without it. but otherwise it's a pretty clean cutoff.

File: kube-client/src/lib.rs
Comment (clux) [reply kube-client/src/lib.rs:656]: hm, maybe. the ephemeral container exits pretty quickly right? if that's the case it will likely fix itself for users running integration tests locally - and it doesn't really matter on CI.

File: kube-client/src/lib.rs
Comment (clux) [inline kube-client/src/lib.rs:686]: you could maybe rely on [PartialEq for EphemeralContainer](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.EphemeralContainer.html#impl-PartialEq%3CEphemeralContainer%3E-for-EphemeralContainer) here and do
```suggestion
        assert_eq!(current_containers[0], busybox_eph);
```

Comment (clux) [reply kube-client/src/lib.rs:686]: oh, nm, i see you say some fields get defaulted in the cluster

File: kube-client/src/lib.rs
Comment (clux) [reply kube-client/src/lib.rs]: tests are much cleaner now :-)

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs]: > I took the liberty of updating the other sub resource documentation examples as well.

appreciate it!

> Might have hidden too many imports in the examples though 🤔

eh, yeah maybe. i would have left the `k8s_openapi` and `kube` import but it's not a big deal.

File: kube-client/src/lib.rs
Comment (clux) [inline kube-client/src/lib.rs:712]: don't they go away quickly? your sleep time is very short

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs:249]: > perhaps Kubernetes ignoring all of the fields that do not belong to the resource being patched.

mm, that's what i am thinking as well.

> I am only worried about the lack of errors when submitting an invalid patch, hence the examples in the documentation.

yeah, i definitely want to have users having clear info about when strange things can fail as well. just wanted to highlight two separate concerns here;

1. certain patch types seem to need typemeta included in the patch as a general thing
2. certain api endpoints not giving you back errors for invalid patches.

it is useful to highlight awkward patch behaviour under the ephemeral resources, but we should document general findings under `Patch` docs so people can more easily notice it. we should possibly even extend `PatchParams::validate` for this if it's a footgun (we can make a separate issue for this though).

if the broken api endpoint is just the ephemeral container one, (e.g. if it cannot be reproduced via straight `Api::patch`), then it makes sense to have an extra warning for these methods. it also might be worth raising a bug upstream against kubernetes long term.

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs]: Future note: the choice of having all signatures require a `K` still feels sensible to me, but dependent on how portable sending a full K is everywhere. our suspicion here is the apiserver just discards everything not relevant, in which case great. Pod is fully optional so it's not harder for users to pass a K, so we should keep `K` for ephemeral. Coercing to a `K` is a nice way to get type-safety and it means passing less vectors around. BUT whether it's generally better (for all subresources) depends on whether we have [apply-configurations](https://github.com/kube-rs/kube/issues/649#issuecomment-939987209) for all structs. 

If is actually slightly problematic for `CRDs` atm because they hardcode `.spec` as non-optional. It only applies to the general `patch_status` call (which uses a vector rather than a `K`) because `patch_scale` sidesteps the issue entirely by having a separate [`Scale`](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/autoscaling/v1/struct.Scale.html) struct. If we made `patch_status` take a `K`, then we would need to make `spec` optional because otherwise it'd be impossible to create patches for it (without requiring `Default` on spec or something else wacky).

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: roger that, I have left them visible now.

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: That all sounds sensible to me. Just to confirm, are we leaving  `replace_ephemeral_containers` as it is in the current version? meaning it accepts a `K: serializable`

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs:249]: > 
> if the broken api endpoint is just the ephemeral container one, (e.g. if it cannot be reproduced via straight `Api::patch`), then it makes sense to have an extra warning for these methods. it also might be worth raising a bug upstream against kubernetes long term.

I just tried this with `Api::patch` and I get the same "bad" behavior :(. I will move the comments regarding this to the general `Patch` docs and make a separate issue to discuss  extending `PatchParams::validate`.

> certain patch types seem to need typemeta included in the patch as a general thing

Ah I see, is there anything I should address related to this?

Comment (jmintb) [reply kube-client/src/api/subresource.rs:249]: I might have overdone the `Patch` docs haha, let me know what you think.

File: kube-client/src/lib.rs
Comment (jmintb) [reply kube-client/src/lib.rs:656]: They do exit quickly, but the Pod is not cleaned up. It will just stay as completed, requiring manual cleanup.

I could just add a "delete pod if it already exists" to the start of the test. :thinking:

File: kube-client/src/lib.rs
Comment (jmintb) [reply kube-client/src/lib.rs:712]: They still persist in the pod spec :(

File: kube-client/src/api/subresource.rs
Comment (jmintb) [reply kube-client/src/api/subresource.rs]: `PatchParams::validate_strict` seems to already do what we need to cover for this case. I tried it out and it does reject the invalid patch example :partying_face:

**Comment by jmintb** [comment]
The cargo deny lint is producing peculiar errors :thinking: , not sure what is causing this.

```

error[duplicate]: found 2 duplicate entries for crate 'syn'
    ┌─ /github/workspace/Cargo.lock:170:1
    │  
170 │ ╭ syn 1.0.109 registry+https://github.com/rust-lang/crates.io-index
171 │ │ syn 2.0.0 registry+https://github.com/rust-lang/crates.io-index
    │ ╰───────────────────────────────────────────────────────────────^ lock entries
```

```
error[duplicate]: found 2 duplicate entries for crate 'hermit-abi'
   ┌─ /github/workspace/Cargo.lock:65:1
   │  
65 │ ╭ hermit-abi 0.2.6 registry+https://github.com/rust-lang/crates.io-index
66 │ │ hermit-abi 0.3.1 registry+https://github.com/rust-lang/crates.io-index
   │ ╰──────────────────────────────────────────────────────────────────────^ lock entries
   
  ```

File: kube-client/src/lib.rs
Comment (jmintb) [inline kube-client/src/lib.rs:615]: This does not seem to work :thinking: , v1.21 integration tests are still failing due to this test.

**Comment by jmintb** [comment]
Just checking in, is there anything I am missing before the next round of review? :) 

If you are just busy then there is no rush :+1: , just wanted to make sure that you weren't waiting for me to do something.

I will have another look at the failing CI and versioning guarding the tests.

**Comment by clux** [comment]
Ah, sorry, I have been a bit busy. Tons of PRs for 0.81 (which i plan on releasing today).
There's been some changes to parameter structs so might be useful to update the branch at this point.

File: kube-client/src/lib.rs
Comment (clux) [reply kube-client/src/lib.rs:615]: huh. that's odd. i'm not sure i fully grok [the macro](https://docs.rs/k8s-openapi/latest/src/k8s_openapi/opt/rustwide/target/x86_64-unknown-linux-gnu/debug/build/k8s-openapi-5fdad33d0cb69d91/out/conditional_compilation_macros.rs.html#86), but thought for sure this would not get run with that..

maybe you have to move it to the bottom in a top level `k8s_if_ge_1_25! { #[cfg(test)] mod test {...} }`. might be able to test this locally with whatever thing you have running locally by varying the version to match your local cluster (if you have one)

File: kube-client/src/api/subresource.rs
Comment (clux) [reply kube-client/src/api/subresource.rs]: Yeah, let's leave this as-is for now. I'll resolve this thread after a follow-up issue.

**Comment by clux** [comment]
> The cargo deny lint is producing peculiar errors thinking , not sure what is causing this.

Don't worry about that. I expect it will be gone after an update, but it can be resolved separately. The deny dependency duplication is a bit annoying for CI since it can break randomly - but it is useful info for us.

File: kube-client/src/lib.rs
Comment (jmintb) [reply kube-client/src/lib.rs:614]: The macro actually seems to be working. The issue is that it looks at the features enabled for `k8s-openapi` which has nothing to do with the API version of the cluster. [this](https://github.com/jmintb/kube-rs/blob/82512c732dcd80325ca3a7214f2fa8af62ba4f1c/kube-client/Cargo.toml#L88) line enables the `v1_26` feature during tests, which intern causes the macro to include the ephemeral containers test.

I am not sure what would be the best way to get around this, perhaps a check inside the test for the actual cluster version? seems a bit hacky though.

Comment (clux) [reply kube-client/src/lib.rs:614]: Ah, damn. I guess we've never have tests that actually had tests that target features like this. In the past I think we've stuffed them in examples and only run them as a one-off for the version.

..you could possibly use [`Client::apiserver_version`](https://docs.rs/kube/latest/kube/struct.Client.html#method.apiserver_version), the result struct contains a [major and a minor](https://docs.rs/k8s-openapi/0.18.0/k8s_openapi/apimachinery/pkg/version/struct.Info.html). I think that's actually sensible.

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:474]: This is non-serverside style patch that will not work without `metadata.name` and typemeta (aka a complete `Pod` struct equivalent).

The recommendation here is to use `validation_strict` only makes sense to me when we are doing serverside apply. If this is a failure of `Strategic` or `Merge` then it's strange that it works (as in bubbles up an error?) when we use `:;validation_strict()` :thinking:

File: kube-client/src/lib.rs
Comment (jmintb) [reply kube-client/src/lib.rs:614]: Awesome! I will give that shot.

**Comment by jmintb** [comment]
I had to focus on uni for a bit. Will get back on this now.

File: kube-core/src/params.rs
Comment (jmintb) [reply kube-core/src/params.rs:509]: Ah yes,  I see what you mean :facepalm: I will go over this again.

Comment (jmintb) [reply kube-core/src/params.rs:509]: After testing again non-server side patches do appear to be affected by the strict validation. The error returned does not appear to originate from field validation but rather that the cluster can not process the incorrect(which is correct). Maybe this is some side effect of enabling strict validation. I can not find anything indidicating this behavior in the [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#validation-for-unrecognized-or-duplicate-fields-setting-the-field-validation-level).

I don't think investigating this behavior is not relevant for this PR, so I will dump my findings into a separate issue.

```rust
        let invalid_patch: PodSpec = serde_json::from_value(serde_json::json!({
                        "activeDeadlineSeconds": 5
        }))?;

        pods.patch(
            &pod_name,
            &PatchParams::default().validation_strict(),
            &Patch::Strategic(invalid_patch),
        )
        .await?;

```

I get the following error:

```
Error: Api(ErrorResponse { status: "Failure", message: " \"\" is invalid: patch: Invalid value: \"map[activeDeadlineSeconds:5 containers:[]]\": strict decoding error: unknown field \"activeDeadlineSeconds\", unknown field \"containers\"", reason: "Invalid", code: 422 })
```

Comment (jmintb) [reply kube-core/src/params.rs]: I think the best thing for the example in the docs is to just use serve-side patches. That also is inline with the general recommendation presented in the docs as well.

Comment (jmintb) [reply kube-core/src/params.rs]: I have applied these changes and attempted to clean up the formulation a bit. Let me know what you think :)

Comment (jmintb) [reply kube-core/src/params.rs]: > After testing again non-server side patches do appear to be affected by the strict validation. The error returned does not appear to originate from field validation. The 422 indicates that the cluster can not process the incorrect object(which is accurate). Maybe this is some side effect of enabling strict validation. I can not find anything indicating this behavior in the [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#validation-for-unrecognized-or-duplicate-fields-setting-the-field-validation-level).
> 
> I don't think investigating this behavior is relevant for this PR, I will dump my findings into a separate issue for later. Feel free to object.
> 
> ```rust
>         let invalid_patch: PodSpec = serde_json::from_value(serde_json::json!({
>                         "activeDeadlineSeconds": 5
>         }))?;
> 
>         pods.patch(
>             &pod_name,
>             &PatchParams::default().validation_strict(),
>             &Patch::Strategic(invalid_patch),
>         )
>         .await?;
> ```
> 
> I get the following error:
> 
> ```
> Error: Api(ErrorResponse { status: "Failure", message: " \"\" is invalid: patch: Invalid value: \"map[activeDeadlineSeconds:5 containers:[]]\": strict decoding error: unknown field \"activeDeadlineSeconds\", unknown field \"containers\"", reason: "Invalid", code: 422 })
> ```
 
 
Follow up, performing the same test with server-side patch returns a 400 complaining about patch version. This is more inline with the docs I think. As the strict validation is supposed to return a 400.

```
Api(ErrorResponse { status: "Failure", message: "Incorrect version specified in apply patch. Specified patch version: , expected: v1", reason: "BadRequest", code: 400 })
```

Comment (jmintb) [reply kube-core/src/params.rs:506]: > I think the best thing for the example in the docs is to just use serve-side patches. That also is inline with the general recommendation presented in the docs as well.

Sorry for dragging this out :sweat: 

Using server-side patching mitigates the issue of incorrect patches, as the patch is rejected due to a missing patch version. Therefore I have reverted the example to use a strategic patch and noted that serve-side patches avoid this issue.

> This is non-serverside style patch that will not work without `metadata.name` and typemeta (aka a complete `Pod` struct equivalent).

This does not appear to be case :thinking: unless kube-rs automatically adds `metadata.name` I can perform non server-side patches just fine without.

File: kube-client/src/lib.rs
Comment (clux) [reply kube-client/src/lib.rs:671]: I see this is done and documented well now :+1:

File: kube-core/src/params.rs
Comment (clux) [inline kube-core/src/params.rs:500]: As a future note to self: I am guessing this happens because `activeDeadlineSeconds` is not a field on the top level struct expected in the patch on the go side, and thus once it's deserialized, it's the apiserver is simply asked to do an empty patch, which should never be an error.

File: kube-core/src/params.rs
Comment (clux) [reply kube-core/src/params.rs:506]: > > This is non-serverside style patch that will not work without `metadata.name` and typemeta (aka a complete `Pod` struct equivalent).
> 
> This does not appear to be case thinking unless kube-rs automatically adds `metadata.name`. I can perform non server-side patches just fine without.

 Kube does not add data to patches anywhere. Maybe I am misremembering this entirely. Looking over this now I see `pod_api` example doing a `PatchParams::default` with a `Patch::Merge` without a `metadata.name`.

I am unable to verify this right now due to a desktop issue, but at any rate, we do not need to block the PR over this. You have sorted out everything for the subresource, the `Api` impl, and everything seems good on CI. 

How do you feel about picking out the `params.rs` docs changes into its own PR? That is the only thing that I feel warrants one extra verification. Am happy to merge the remaining changes.

File: kube-core/src/params.rs
Comment (jmintb) [reply kube-core/src/params.rs:500]: Exactly!

File: kube-core/src/params.rs
Comment (jmintb) [reply kube-core/src/params.rs:506]: That sounds perfect. `params.rs` needs a second pair of eyes.

Comment (jmintb) [reply kube-core/src/params.rs]: Done: https://github.com/kube-rs/kube/pull/1262

**Review by clux — APPROVED** [review]
Looks great. Thank you so much. Appreciate the perseverance through the amount of annoyingly complex details that arose in here.

**Pull request merged**
